src/realm/object-store/sectioned_results.cpp

```diff
@@ -0,0 +1,385 @@
+                                           section_changes.first, section_changes.second},
+                 {});
+        }
+
+        REALM_ASSERT(m_sectioned_results.m_results.is_valid());
+
+        m_prev_row_to_index_path = m_sectioned_results.m_row_to_index_path;
+        m_prev_sections = m_sectioned_results.m_sections;
+    }
+    void error(std::exception_ptr ptr)
+    {
+        m_cb({}, ptr);
+    }
+
+    std::pair<IndexSet, IndexSet> calculate_sections_to_insert_and_delete()
+    {
+        IndexSet sections_to_insert, sections_to_remove;
+        std::map<size_t, size_t> new_sections, old_sections;
+
+        for (auto& n : m_sectioned_results.m_sections) {
+            new_sections[n.hash] = n.index;
+        }
+        for (auto& n : m_prev_sections) {
+            old_sections[n.hash] = n.index;
+        }
+
+        auto diff = [](const std::map<size_t, size_t>& a, const std::map<size_t, size_t>& b, IndexSet& out) {
+            for (auto& [hash, index] : b) {
+                if (a.find(hash) == a.end()) {
+                    out.add(b.at(hash));
+                }
+            }
+        };
+
+        diff(old_sections, new_sections, sections_to_insert);
+        diff(new_sections, old_sections, sections_to_remove);
+
+        return {sections_to_insert, sections_to_remove};
+    }
+
+private:
+    SectionedResultsNotificatonCallback m_cb;
+    SectionedResults& m_sectioned_results;
+    std::vector<Section> m_prev_sections;
+    std::vector<std::pair<size_t, size_t>> m_prev_row_to_index_path;
+    util::Optional<size_t> m_section_filter;
+};
+
+Mixed ResultsSection::operator[](size_t idx) const
+{
+    m_parent->calculate_sections_if_required();
+    return m_parent->m_results.get_any(m_parent->m_sections[m_index].indices[idx]);
+}
+
+Mixed ResultsSection::key()
+{
+    return m_parent->m_sections[m_index].key;
+}
+
+size_t ResultsSection::size()
+{
+    m_parent->calculate_sections_if_required();
+    REALM_ASSERT(m_parent->m_sections.size() > m_index);
+    auto range = m_parent->m_sections[m_index];
+    return range.indices.size();
+}
+
+NotificationToken ResultsSection::add_notification_callback(SectionedResultsNotificatonCallback callback,
+                                                            KeyPathArray key_path_array) &
+{
+    return m_parent->add_notification_callback_for_section(m_index, std::move(callback), key_path_array);
+}
+
+realm::ThreadSafeReference ResultsSection::thread_safe_reference()
+{
+    return m_parent->thread_safe_reference();
+}
+
+bool ResultsSection::is_valid() const
+{
+    return m_parent->is_valid();
+}
+
+size_t ResultsSection::hash() const
+{
+    return m_parent->m_sections[m_index].hash;
+}
+
+SectionedResults::SectionedResults(Results results, SectionKeyFunc section_key_func)
+    : m_results(results)
+    , m_callback(std::move(section_key_func))
+{
+    calculate_sections_if_required(true);
```

Comment:
We don't want to force eager evaluation of the sections when the SectionedResults are created. Results being lazy is required for async queries which never run the query/sort on the main thread to work.